msrestazure/azure_local_creds_prober.py

```diff
@@ -0,0 +1,142 @@
+                creds = ServicePrincipalCredentials(client_id=auth_info['clientId'],
+                                                    secret=auth_info['clientSecret'],
+                                                    tennt_id=auth_info['tenantId'])
+            elif os.environ.get('AZURE_CLIENT_ID'):
+                # TODO error out if other env vars are not set
+                creds = ServicePrincipalCredentials(client_id=os.environ.get('AZURE_CLIENT_ID'),
+                                                    secret=os.environ.get('AZURE_CLIENT_SECRET'),
+                                                    tennt_id=os.environ.get('AZURE_TENANT_ID'))
+                _LOGGER.warning('Service principal credentials was detected')
+            else:
+                try:
+                    creds = CLICredentials()
+                    _LOGGER.warning('Azure CLI credentials was detected')
+                except NotImplementedError:
+                    raise ValueError('No credential was detected from the local machine')
+        self.creds = creds
+        if not subscription_id:
+            try:
+                from azure.mgmt.resource.subscriptions import SubscriptionClient
+                subscriptions = list(SubscriptionClient(creds).subscriptions.list())
+                if subscriptions:
+                    subscription_ids = [s.id.split('/')[-1] for s in subscriptions]
+                    subscription_id = subscription_ids[0]
+                    _LOGGER.warning('Found subscription "%s" to use', subscription_ids[0])
+                    if len(subscription_ids) > 1:
+                        _LOGGER.warning('You also have accesses to a few other subscriptions "%S".'
+                                        ' You can supply subscription_id on creating the probe object')
+            except ImportError:  # should be rare
+                _LOGGER.warning('Failed to load azure.mgmt.resource.subscriptions to find the default subscription.'
+                                ' If this is expected, supply subscription_id on creating the probe object')
+        self.subscription_id = subscription_id
+
+
+class CLICredentials(BasicTokenAuthentication):
+
+    def __init__(self, resource=None, subscription_id=None):   # allow subscriptions
+        super(CLICredentials, self).__init__(None)
+        uname = platform.uname()
+        # python 2, `platform.uname()` returns: tuple(system, node, release, version, machine, processor)
+        platform_name = getattr(uname, 'system', None) or uname[0]
+        platform_name = platform_name.lower()
+        if platform_name == 'windows':
+            program_files_folder = os.environ.get('ProgramFiles(x86)') or os.environ.get('ProgramFiles')
+            probing_paths = [os.path.join(program_files_folder, 'Microsoft SDKs', 'Azure', 'CLI2', 'wbin', 'az.cmd')]
+        else:
+            probing_paths = ['/usr/bin/az', '/usr/local/bin/az']
```

Comment:
This won't work if CLI was installed by pip. You could subprocess "which az", or us "shutil.which('az')" starting Python 3
https://docs.python.org/3/library/shutil.html?highlight=.which#shutil.which

Reply:
Can we use the presence of a ${HOME}/.azure file as the hint to try the CLI creds?

Reply:
...or, alternatively, use the ${PATH} environment variable...